ec2_terminator/app.py

```diff
@@ -0,0 +1,157 @@
+import json
+import logging
+import os
+import time
+
+import boto3
+
+
+logging.basicConfig(level=logging.DEBUG,
+                    format='%(asctime)s %(levelname)s %(message)s')
+for lib in ["botocore", "urllib3"]:
+    log = logging.getLogger(lib)
+    log.setLevel(logging.WARNING)
+
+
+def list_regions():
+    client = boto3.client('ec2')
+
+    region_desc = client.describe_regions()
+    regions = [region['RegionName'] for region in region_desc['Regions']]
+
+    logging.debug(f"Regions: {regions}")
+    return regions
+
+
+def list_instances(region):
+    """Return a list of instance IDs in the given region to stop or terminate"""
+    client = boto3.client('ec2', region_name=region)
+
+    # This intentionally reprocesses stopped instances for debugging
+    state_filters = [{
+        'Name': 'instance-state-name',
+        'Values': ['running', 'stopping', 'stopped'],
+    }]
+
+    instances = []
+
+    pager = client.get_paginator('describe_instances')
+    for page in pager.paginate(Filters=state_filters):
+        for rsvp in page['Reservations']:
+            for ec2 in rsvp['Instances']:
+                ec2_id = ec2['InstanceId']
+                logging.debug(f"EC2: {ec2}")
+                instances.append(ec2_id)
+
+    logging.debug(f"Instances found: {instances}")
+    return instances
+
+
+def stop_instances(instances, region):
```

Comment:
couldn't this function and `terminate_instances` be combined into a general `set_instance_state(instances, state, region)` function?  Where the state is their `stop` or `terminate`?

Reply:
I took a stab at implementing this, but there isn't a similar boto3 function, so there's still two blocks of code for the two different cases of calling `client.stop_instances()` or `client.terminate_instances` and then parsing the output for either `StoppingInstances` or `TerminatingInstances`. This does make it easier to add other cases, but the only other similar boto3 functions are `client.start_instances`, `client.reboot_instances`, `client.monitor_instances`, and `client.unmonitor_instances`. I guess I can add them to make this more of a general lifecycle manager, what are your thoughts on the usefulness of those functions?

Reply:
the other lifecycle options are probably not useful for us currently however setting it up to allow us to easily add them is useful.  I'm thinking either someone else will find them useful or that AWS will add additional options that will be useful for other use cases.